In [37]:
# Week 8: Text Pre-processing and Text Classification
# Import Required Libraries

import pandas as pd
import numpy as np
import re

In [38]:
# Mount Google Drive

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
# Set Dataset Path

base_path = "/content/drive/MyDrive/Week-8"

print(base_path)

/content/drive/MyDrive/Week-8


In [40]:
#Load the Pre-processing Dataset

preprocessing_file = base_path + "/trumptweets_small.csv"

df = pd.read_csv(preprocessing_file)

df.head()

,id,link,content,date,retweets,favorites,mentions,hashtags,geo
0,1698308935,https://twitter.com/realDonaldTrump/status/169...,Be sure to tune in and watch Donald Trump on L...,2009-05-04 20:54:25,500,868,NaN,NaN,NaN
1,1701461182,https://twitter.com/realDonaldTrump/status/170...,Donald Trump will be appearing on The View tom...,2009-05-05 03:00:10,33,273,NaN,NaN,NaN
2,1737479987,https://twitter.com/realDonaldTrump/status/173...,Donald Trump reads Top Ten Financial Tips on L...,2009-05-08 15:38:08,12,18,NaN,NaN,NaN
3,1741160716,https://twitter.com/realDonaldTrump/status/174...,New Blog Post: Celebrity Apprentice Finale and...,2009-05-08 22:40:15,11,24,NaN,NaN,NaN
4,1773561338,https://twitter.com/realDonaldTrump/status/177...,"""My persona will never be that of a wallflower...",2009-05-12 16:07:28,1399,1965,NaN,NaN,NaN


In [41]:
# Check Dataset Information

print("Dataset shape:", df.shape)

print("\nDataset columns:")
print(df.columns)

print("\nMissing values:")
print(df.isnull().sum())

Dataset shape: (41122, 9)

Dataset columns:
Index(['id', 'link', 'content', 'date', 'retweets', 'favorites', 'mentions',
       'hashtags', 'geo'],
      dtype='object')

Missing values:
id               0
link             0
content          0
date             0
retweets         0
favorites        0
mentions     18655
hashtags     35312
geo          41122
dtype: int64


In [42]:
# Select Text Column for Pre-processing

df_text = df[["content"]]

df_text.head()

,content
0,Be sure to tune in and watch Donald Trump on L...
1,Donald Trump will be appearing on The View tom...
2,Donald Trump reads Top Ten Financial Tips on L...
3,New Blog Post: Celebrity Apprentice Finale and...
4,"""My persona will never be that of a wallflower..."


## Remove URLs

In [43]:
# Remove URLs from Text

def remove_urls(text):
    """
    Remove URLs from a text string.
    """
    url_pattern = re.compile(r"https?://\S+|www\.\S+")
    return url_pattern.sub("", text)

Test the function

In [44]:
# Test remove_urls function

sample_text = "Click here to open Instagram https://instagram.com"

text_no_url = remove_urls(sample_text)

print(text_no_url)

Click here to open Instagram 


Apply URL Removal to Dataset

In [45]:
# Apply URL Removal to Dataset

text_no_url = df_text["content"].apply(remove_urls)

text_no_url.head()

,content
0,Be sure to tune in and watch Donald Trump on L...
1,Donald Trump will be appearing on The View tom...
2,Donald Trump reads Top Ten Financial Tips on L...
3,New Blog Post: Celebrity Apprentice Finale and...
4,"""My persona will never be that of a wallflower..."


## Remove Emojis

In [46]:
# Remove Emojis from Text

def remove_emoji(text):
    """
    Remove emojis from a text string.
    """
    emoji_pattern = re.compile("["
                               u"\U0001F600-\U0001F64F"  # emoticons
                               u"\U0001F300-\U0001F5FF"  # symbols and pictographs
                               u"\U0001F680-\U0001F6FF"  # transport and map symbols
                               u"\U0001F1E0-\U0001F1FF"  # flags
                               u"\U00002702-\U000027B0"
                               u"\U000024C2-\U0001F251"
                               "]+", flags=re.UNICODE)

    return emoji_pattern.sub(" ", text)

Test the function

In [47]:
# Test remove_emoji function

sample_text = "Hello everyone 👋😊 this tweet is exciting 🔥"

text_no_emoji = remove_emoji(sample_text)

print(text_no_emoji)

Hello everyone   this tweet is exciting  


Apply Emoji Removal to Dataset

In [48]:
# Apply Emoji Removal to Dataset

text_no_emoji = text_no_url.apply(remove_emoji)

text_no_emoji.head()

,content
0,Be sure to tune in and watch Donald Trump on L...
1,Donald Trump will be appearing on The View tom...
2,Donald Trump reads Top Ten Financial Tips on L...
3,New Blog Post: Celebrity Apprentice Finale and...
4,"""My persona will never be that of a wallflower..."


Remove Mentions, Hashtags, and Special Characters

In [49]:
# Remove Mentions, Hashtags, and Special Characters

def remove_unwanted_characters(text):
    """
    Remove mentions, hashtags, punctuation, and special characters from text.
    """
    text = re.sub(r"@[A-Za-z0-9_]+", " ", text)      # remove mentions
    text = re.sub(r"#[A-Za-z0-9_]+", " ", text)      # remove hashtags
    text = re.sub(r"[^A-Za-z0-9]+", " ", text)      # remove punctuation/special characters
    text = re.sub(r"\s+", " ", text)                # remove extra spaces

    return text.strip()

Test the function

In [50]:
# Test remove_unwanted_characters function

sample_text = "Hello @student! This is #Week8 NLP task!!!"

cleaned_text = remove_unwanted_characters(sample_text)

print(cleaned_text)

Hello This is NLP task


Apply Unwanted Character Removal to Dataset

In [51]:
# Apply Unwanted Character Removal to Dataset

text_cleaned = text_no_emoji.apply(remove_unwanted_characters)

text_cleaned.head()

,content
0,Be sure to tune in and watch Donald Trump on L...
1,Donald Trump will be appearing on The View tom...
2,Donald Trump reads Top Ten Financial Tips on L...
3,New Blog Post Celebrity Apprentice Finale and ...
4,My persona will never be that of a wallflower ...


Convert Text to Lowercase

In [52]:
# Convert Text to Lowercase

def lower_order(text):
    """
    Convert text into lowercase.
    """
    return text.lower()

Test the function

In [53]:
# Test lower_order function

sample_text = "This Is Some Normalized TEXT"

sample_lower = lower_order(sample_text)

print(sample_lower)


this is some normalized text


Apply Lowercase Conversion to Dataset

In [54]:
# Apply Lowercase Conversion to Dataset

text_lower = text_cleaned.apply(lower_order)

text_lower.head()

,content
0,be sure to tune in and watch donald trump on l...
1,donald trump will be appearing on the view tom...
2,donald trump reads top ten financial tips on l...
3,new blog post celebrity apprentice finale and ...
4,my persona will never be that of a wallflower ...


In [90]:
# Import NLTK libraries for tokenization

import nltk
from nltk import word_tokenize
from nltk.tokenize import RegexpTokenizer

## Tokenization

In [55]:
# Download required tokenizer resource

nltk.download("punkt")
nltk.download("punkt_tab")
# Download stopwords list



[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [56]:
# Tokenization

# Test case
input_text = "He did not try to navigate after the first bold flight, for the reaction had taken something out of his soul."

tokenized_text = word_tokenize(input_text)

tokenized_text

['He',
 'did',
 'not',
 'try',
 'to',
 'navigate',
 'after',
 'the',
 'first',
 'bold',
 'flight',
 ',',
 'for',
 'the',
 'reaction',
 'had',
 'taken',
 'something',
 'out',
 'of',
 'his',
 'soul',
 '.']

Remove Punctuation

In [57]:
# Remove Punctuation

def remove_punct(text_tokens):
    """
    Remove punctuation from tokenized text.
    """
    tokenizer = RegexpTokenizer(r"\w+")
    clean_tokens = tokenizer.tokenize(" ".join(text_tokens))

    return clean_tokens

Test the function

In [58]:
# Test remove_punct function

text_punctuation = "He did not try to navigate: after the!!!! first bold flight, for,,,,, the reaction!!!!had taken??????? something out of his soul."

text_punctuation_tokens = word_tokenize(text_punctuation)

print(text_punctuation_tokens)

clean_tokens = remove_punct(text_punctuation_tokens)

print(clean_tokens)

['He', 'did', 'not', 'try', 'to', 'navigate', ':', 'after', 'the', '!', '!', '!', '!', 'first', 'bold', 'flight', ',', 'for', ',', ',', ',', ',', ',', 'the', 'reaction', '!', '!', '!', '!', 'had', 'taken', '?', '?', '?', '?', '?', '?', '?', 'something', 'out', 'of', 'his', 'soul', '.']
['He', 'did', 'not', 'try', 'to', 'navigate', 'after', 'the', 'first', 'bold', 'flight', 'for', 'the', 'reaction', 'had', 'taken', 'something', 'out', 'of', 'his', 'soul']


Apply Tokenization and Punctuation Removal to Dataset

In [59]:
# Apply Tokenization and Punctuation Removal to Dataset

tokens = text_lower.apply(word_tokenize)

tokens_no_punct = tokens.apply(remove_punct)

tokens_no_punct.head()

,content
0,"[be, sure, to, tune, in, and, watch, donald, t..."
1,"[donald, trump, will, be, appearing, on, the, ..."
2,"[donald, trump, reads, top, ten, financial, ti..."
3,"[new, blog, post, celebrity, apprentice, final..."
4,"[my, persona, will, never, be, that, of, a, wa..."


Remove Stopwords

In [91]:
# Import stopwords library

from nltk.corpus import stopwords

In [60]:
# Remove Stopwords

nltk.download("stopwords")

stop_words = set(stopwords.words("english"))

custom_stopwords = ["@", "RT", "rt"]
stop_words.update(custom_stopwords)

def remove_stopwords(text_tokens):
    """
    Remove common English stopwords from tokenized text.
    """
    result_tokens = []

    for token in text_tokens:
        if token not in stop_words:
            result_tokens.append(token)

    return result_tokens

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Test the function

In [61]:
# Test remove_stopwords function

test_tokens = [
    "He", "did", "not", "try", "to", "navigate", "after",
    "the", "first", "bold", "flight", "for", "the",
    "reaction", "had", "taken", "something", "out", "of",
    "his", "soul"
]

tokens_without_stopwords_test = remove_stopwords(test_tokens)

print(test_tokens)
print(tokens_without_stopwords_test)

['He', 'did', 'not', 'try', 'to', 'navigate', 'after', 'the', 'first', 'bold', 'flight', 'for', 'the', 'reaction', 'had', 'taken', 'something', 'out', 'of', 'his', 'soul']
['He', 'try', 'navigate', 'first', 'bold', 'flight', 'reaction', 'taken', 'something', 'soul']


Apply Stopword Removal to Dataset

In [62]:
# Apply stopword removal to dataset

tokens_without_stopwords = tokens_no_punct.apply(remove_stopwords)

tokens_without_stopwords.head()

,content
0,"[sure, tune, watch, donald, trump, late, night..."
1,"[donald, trump, appearing, view, tomorrow, mor..."
2,"[donald, trump, reads, top, ten, financial, ti..."
3,"[new, blog, post, celebrity, apprentice, final..."
4,"[persona, never, wallflower, rather, build, wa..."


In [92]:
# Import text normalization libraries

from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer

## Text Normalization

Text normalization reduces different forms of words into a common form.  
In this section, lemmatization and stemming are used to reduce word variations.

Lemmatization

In [63]:
# Lemmatization

nltk.download("wordnet")
nltk.download("omw-1.4")

lemmatizer = WordNetLemmatizer()

def lemmatization(token_text):
    """
    Convert tokens into their base form using lemmatization.
    """
    lemmatized_tokens = []

    for token in token_text:
        lemmatized_tokens.append(lemmatizer.lemmatize(token.lower(), pos="v"))

    return lemmatized_tokens

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


Test the function

In [64]:
# Test lemmatization function

test_tokens = ["Connects", "Connecting", "Connections", "Connected", "Connection", "Connectings", "Connect"]

lemmatized_tokens = lemmatization(test_tokens)

print(test_tokens)
print(lemmatized_tokens)

['Connects', 'Connecting', 'Connections', 'Connected', 'Connection', 'Connectings', 'Connect']
['connect', 'connect', 'connections', 'connect', 'connection', 'connectings', 'connect']


Stemming

In [65]:
# Stemming

stemmer = PorterStemmer()

def stemming(token_text):
    """
    Reduce tokens to their stem form.
    """
    stemmed_tokens = []

    for token in token_text:
        stemmed_tokens.append(stemmer.stem(token))

    return stemmed_tokens

Test the function

In [66]:
# Test stemming function

test_tokens = ["Connects", "Connecting", "Connections", "Connected", "Connection", "Connectings", "Connect"]

stemmed_tokens = stemming(test_tokens)

print(test_tokens)
print(stemmed_tokens)

['Connects', 'Connecting', 'Connections', 'Connected', 'Connection', 'Connectings', 'Connect']
['connect', 'connect', 'connect', 'connect', 'connect', 'connect', 'connect']


Lowercase Conversion

In [67]:
# Lowercase conversion

def lower_order(text):
    """
    Convert text into lowercase.
    """
    return text.lower()

Test the function

In [68]:
# Test lowercase conversion function

sample_text = "This Is some Normalized TEXT"

sample_lower = lower_order(sample_text)

print(sample_lower)

this is some normalized text


Apply Normalization to Dataset

In [69]:
# Apply lemmatization to dataset

lemmatized_tokens = tokens_without_stopwords.apply(lemmatization)

lemmatized_tokens.head()

,content
0,"[sure, tune, watch, donald, trump, late, night..."
1,"[donald, trump, appear, view, tomorrow, mornin..."
2,"[donald, trump, read, top, ten, financial, tip..."
3,"[new, blog, post, celebrity, apprentice, final..."
4,"[persona, never, wallflower, rather, build, wa..."


Convert Tokens Back to Text

In [70]:
# Convert token lists back into text

cleaned_text = lemmatized_tokens.apply(lambda tokens: " ".join(tokens))

cleaned_text.head()

,content
0,sure tune watch donald trump late night david ...
1,donald trump appear view tomorrow morning disc...
2,donald trump read top ten financial tip late s...
3,new blog post celebrity apprentice finale less...
4,persona never wallflower rather build wall cli...


Text Cleaning Pipeline

In [71]:
# Text cleaning pipeline

def text_cleaning_pipeline(text, rule="lemmatize"):
    """
    Clean text by applying all preprocessing steps.
    """
    text = lower_order(text)
    text = remove_urls(text)
    text = remove_emoji(text)
    text = remove_unwanted_characters(text)

    tokens = word_tokenize(text)
    tokens = remove_punct(tokens)
    tokens = remove_stopwords(tokens)

    if rule == "lemmatize":
        tokens = lemmatization(tokens)
    elif rule == "stem":
        tokens = stemming(tokens)
    else:
        raise ValueError("Rule must be either 'lemmatize' or 'stem'.")

    return " ".join(tokens)

Test the pipeline

In [72]:
# Test text cleaning pipeline

sample_text = "Hello @student 👋, I am studying NLP!!! #Week8 https://example.com"

cleaned_sample = text_cleaning_pipeline(sample_text)

print(cleaned_sample)

hello study nlp


Apply Text Cleaning Pipeline to Dataset

In [73]:
# Apply text cleaning pipeline to dataset

df["cleaned_content"] = df["content"].apply(text_cleaning_pipeline)

df[["content", "cleaned_content"]].head()

,content,cleaned_content
0,Be sure to tune in and watch Donald Trump on L...,sure tune watch donald trump late night david ...
1,Donald Trump will be appearing on The View tom...,donald trump appear view tomorrow morning disc...
2,Donald Trump reads Top Ten Financial Tips on L...,donald trump read top ten financial tip late s...
3,New Blog Post: Celebrity Apprentice Finale and...,new blog post celebrity apprentice finale less...
4,"""My persona will never be that of a wallflower...",persona never wallflower rather build wall cli...


Final Check for Pre-processing

In [74]:
# Check final cleaned data

print("Original text:")
print(df["content"].iloc[0])

print("\nCleaned text:")
print(df["cleaned_content"].iloc[0])


Original text:
Be sure to tune in and watch Donald Trump on Late Night with David Letterman as he presents the Top Ten List tonight!

Cleaned text:
sure tune watch donald trump late night david letterman present top ten list tonight


## Text Classification Using Machine Learning

In [93]:
# Import machine learning libraries

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

Load Classification Dataset

In [75]:
# Load classification dataset

classification_file = base_path + "/trum_tweet_sentiment_analysis.csv"

classification_df = pd.read_csv(classification_file)

classification_df.head()

,text,Sentiment
0,RT @JohnLeguizamo: #trump not draining swamp b...,0
1,ICYMI: Hackers Rig FM Radio Stations To Play A...,0
2,Trump protests: LGBTQ rally in New York https:...,1
3,"""Hi I'm Piers Morgan. David Beckham is awful b...",0
4,RT @GlennFranco68: Tech Firm Suing BuzzFeed fo...,0


Check Classification Dataset Information

In [76]:
# Check classification dataset information

print("Dataset shape:", classification_df.shape)

print("\nDataset columns:")
print(classification_df.columns)

print("\nMissing values:")
print(classification_df.isnull().sum())

print("\nSentiment class distribution:")
print(classification_df["Sentiment"].value_counts())

Dataset shape: (1850123, 2)

Dataset columns:
Index(['text', 'Sentiment'], dtype='object')

Missing values:
text         0
Sentiment    0
dtype: int64

Sentiment class distribution:
Sentiment
0    1244211
1     605912
Name: count, dtype: int64


Apply Text Cleaning Pipeline to Classification Dataset

In [79]:
# Use a smaller sample for faster execution

classification_sample = classification_df.sample(
    n=50000,
    random_state=42
).copy()

classification_sample.shape

(50000, 2)

In [81]:
# Apply text cleaning pipeline to sampled classification dataset

classification_sample["cleaned_text"] = classification_sample["text"].apply(text_cleaning_pipeline)

classification_sample[["text", "cleaned_text", "Sentiment"]].head()

,text,cleaned_text,Sentiment
167359,RT @EthandeMarsi: I don't understand why adult...,understand adults would mean kid age friends w...,0
1495948,RT @atrumptastrophe: Suddenly that 3AM call be...,suddenly 3am call trump flynn last week value ...,0
1592776,@RealJeremyNolt~'going after Trump loyalists'f...,go trump loyalists good reason trump family br...,1
587773,RT @ileriparis: Antoine Sfeir sur #Trump #iler...,antoine sfeir sur,0
1187236,"A United pilot ranted about Trump, Clinton and...",unite pilot rant trump clinton divorce passeng...,0


The cleaning pipeline worked on the 50,000-row sample.

Prepare Features and Target

Now set:

x = cleaned tweet text

y = sentiment label


In [82]:
# Prepare features and target

X = classification_sample["cleaned_text"]
y = classification_sample["Sentiment"]

print("Feature shape:", X.shape)
print("Target shape:", y.shape)

Feature shape: (50000,)
Target shape: (50000,)


Train-Test Split

In [83]:
# Split data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (40000,)
X_test shape: (10000,)
y_train shape: (40000,)
y_test shape: (10000,)


TF-IDF Vectorization

In [84]:
# Convert text into numerical features using TF-IDF

vectorizer = TfidfVectorizer(
    max_features=10000,
    min_df=5,
    max_df=0.95,
    stop_words="english"
)

X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

print("Training TF-IDF shape:", X_train_tfidf.shape)
print("Testing TF-IDF shape:", X_test_tfidf.shape)

Training TF-IDF shape: (40000, 5716)
Testing TF-IDF shape: (10000, 5716)


Train Logistic Regression Model

In [85]:
# Train Logistic Regression model

model = LogisticRegression(
    max_iter=1000,
    random_state=42
)

model.fit(X_train_tfidf, y_train)

print("Model training completed.")

Model training completed.


Make Predictions

In [86]:
# Make predictions on test data

y_pred = model.predict(X_test_tfidf)

print("Predictions completed.")

Predictions completed.


Evaluate the Model

In [87]:
# Evaluate model performance

print("Accuracy:", accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.8715

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.94      0.91      6724
           1       0.86      0.72      0.79      3276

    accuracy                           0.87     10000
   macro avg       0.87      0.83      0.85     10000
weighted avg       0.87      0.87      0.87     10000



## Model Evaluation

- The Logistic Regression model achieved **87.15% accuracy**.
- Class **0** performed better than Class **1**.
- Class 0 recall was **0.94**, so most Class 0 tweets were correctly identified.
- Class 1 recall was **0.72**, so some Class 1 tweets were predicted as Class 0.
- The weighted average F1-score was **0.87**, showing good overall performance.

### Confusion Matrix

In [88]:
# Display confusion matrix

conf_matrix = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[6346  378]
 [ 907 2369]]


## Confusion Matrix Interpretation

- **6346** Class 0 tweets were correctly predicted as Class 0.
- **378** Class 0 tweets were incorrectly predicted as Class 1.
- **907** Class 1 tweets were incorrectly predicted as Class 0.
- **2369** Class 1 tweets were correctly predicted as Class 1.
- The model predicts Class 0 better than Class 1.

## Conclusion

- Text preprocessing helped clean noisy tweet data.
- The cleaned text was converted into numerical features using TF-IDF.
- Logistic Regression was used for sentiment classification.
- The model achieved **87.15% accuracy** on the sampled test data.
- Class 0 was predicted better than Class 1.
- Using the full dataset may improve performance but requires more processing time.